broker/fragment/store_azure.go

```diff
@@ -183,17 +225,48 @@ func (a *azureBackend) azureClient(ep *url.URL) (cfg AzureStoreConfig, client pi
 		return
 	}
 
+	var credentials azblob.Credential
+
 	accountName := os.Getenv("AZURE_ACCOUNT_NAME")
-	accountKey := os.Getenv("AZURE_ACCOUNT_KEY")
-	credentials, err := azblob.NewSharedKeyCredential(accountName, accountKey)
-	if err != nil {
-		return
+	a.endpoint = fmt.Sprintf("https://%s.blob.core.windows.net/", accountName)
+
+	if ep.Scheme == "azure" {
+		accountKey := os.Getenv("AZURE_ACCOUNT_KEY")
+		sharedKeyCred, cred_err := service.NewSharedKeyCredential(accountName, accountKey)
+		if cred_err != nil {
+			return cfg, nil, cred_err
+		}
+		serviceClient, cred_err := service.NewClientWithSharedKeyCredential(a.endpoint, sharedKeyCred, &service.ClientOptions{})
```

Comment:
you're building this so that you can then make a rest request to get a delegation key? kind of circuitous in this case.
Why not just hang onto the credentials and keep the signing code path the same for `azure://` ? As a reader, I'd definitely prefer clearly delineated paths for the azure vs azure-ad cases.

It may also make sense to fetch a delegation key only once, below, when initializing for `azure-ad` and thereafter use it for signing credentials.

Reply:
Not entirely sure what you're saying here. We need something that implements `service.Client`, since that's how we get a `UserDelegationCredential`, which we need to build signed URLs. When we're using shared keys (`azure://`), we  [call `NewClientWithSharedKeyCredential()`](https://github.com/gazette/core/pull/341/files#diff-b58442da8fead03ddc65f71858da8439f9f4b987fc32da9b453d9fba6a066c88R252):, and when we're using service principal auth (`azure-ad://`), we [use `NewClient()`](https://github.com/gazette/core/pull/341/files#diff-b58442da8fead03ddc65f71858da8439f9f4b987fc32da9b453d9fba6a066c88R273) and give it the custom token refresher built by `GetAzureStorageCredential()`.

The idea being in the end, we have something that looks like a `service.Client`, which we can call `GetUserDelegationCredential` on, no matter whether we're using `azure://` or `azure-ad://`.

> It may also make sense to fetch a delegation key only once, below, when initializing for azure-ad and thereafter use it for signing credentials.

Yup, I refactored this out into a function that generates the delegation key when it's about to expire, otherwise returns a cached one.

Reply:
> ... since that's how we get a UserDelegationCredential, which we need to build signed URLs

But we don't need a `UserDelegationCredential` in the `azure://` case, that's what I was getting at. That's needed only for the azure-ad case. Before this PR, we would build the structure for signing, and then simply called `.NewSASQueryParameters(a.credentials)`.

The question i'm asking is: does use of a `UserDelegationCredential` represent an improvement for the `azure://` case? It's downside is that it's an obfuscation of what's actually happening in the signing path, and thus _could_ be reserved for use only in the `azure-ad://` case.

I do see that the structures for signing look _nearly_ identical but are technically different structs in different packages. I'm unclear what that's about, or if one structure is actually implemented in terms of the other or something, and whether that's something we could leverage.

I'm not coming down hard on one way vs the other, I'm just looking for "We considered doing ABC but went with XYZ because ..."

Reply:
> But we don't need a `UserDelegationCredential` in the azure:// case

It's somewhat opaque, but my understanding is that this is a way to create SASs that abstracts over whether your underlying credential is a Shared Key (`azure://`), or a Client Secret (`azure-ad://`). In both scenarios we end up with a `azblob/service.Client` which exposes `GetUserDelegationCredential()`. 

I'm not seeing how it creates any more obfuscation? I was actually pretty happy with how it turned out, as the only logic conditional on whether we have `azure://` or `azure://` is in `azureClient()`, at which point we end up with a `service.Client` either way and no longer have to worry about _how_ we're authenticated, just that we are.